File: leet/blind75/sameTree.js
const { treeGen, TreeNode } = require('../viewLib/forest.js')

function isSameTree(p, q) {
  if(p == null && q == null) {
    return true
  }
  if((p == null && q != null) || (p != null && q == null)) {
    return false
  }
  if(p.val != q.val) {
    return false
  }
  return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
}


let vals = [1,2,1]
let vals2 = [1,1,2]
let nodes = []
let nodes2 = []
for(let val of vals) {
  nodes.push(val == null ? null : new TreeNode(val))
}
for(let val of vals2) {
  nodes2.push(val == null ? null : new TreeNode(val))
}

let tree = treeGen(nodes)
let tree2 = treeGen(nodes2)
// console.log(isSameTree(tree, tree2))
 
module.exports = { isSameTree }

File: leet/blind75/sameTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def isSameTree(p: TreeNode, q: TreeNode) -> bool:
    pStack = [p]
    qStack = [q]
    if p is None and q is None:
        return True
    while len(pStack) > 0 and len(qStack) > 0:
        [pNode, qNode] = [pStack.pop(), qStack.pop()]
        if (pNode is None and qNode is not None) or (pNode is not None and qNode is None):
            return False
        if (pNode.left is None and qNode.left is not None) or (pNode.left is not None and qNode.left is None):
            return False
        if (pNode.right is None and qNode.right is not None) or (pNode.right is not None and qNode.right is None):
            return False
        if pNode is not None and qNode is not None:
            if pNode.val != qNode.val:
                return False
            if pNode.left is not None:
                pStack.append(pNode.left)
            if pNode.right is not None:
                pStack.append(pNode.right)
            if qNode.left is not None:
                qStack.append(qNode.left)
            if qNode.right is not None:
                qStack.append(qNode.right)
    return True

